pkg/k8s/resource/resource.go

```diff
@@ -230,6 +246,12 @@ type resource[T k8sRuntime.Object] struct {
 
 	storePromise  promise.Promise[Store[T]]
 	storeResolver promise.Resolver[Store[T]]
+
+	// meaningful for releasable resources only
+	refsMu      lock.Mutex
```

Comment:
Why can't `mu` be used?

Reply:
Initially I did that, but I gave up after trying to solve a race when a resource is released.
During reset we want to block `Events()` and `Store()` until the resource is ready again to be marked as needed. But we also want to wait for all previous goroutines to terminate (and we do that waiting on `r.wg`). Since the `Process` callback used by the informer takes the `r.mu` lock too, it ended up deadlocking.
I tried some alternative implementations like temporary releasing the `r.mu` lock while waiting on `r.wg`, but the code looked way too complicated and it was not able to solve additional corner cases when the resource is also concurrently stopped.